include/envoy/network/connection.h

```diff
@@ -65,8 +65,10 @@ class ConnectionCallbacks {
 enum class ConnectionCloseType {
   FlushWrite, // Flush pending write data before raising ConnectionEvent::LocalClose
   NoFlush,    // Do not flush any pending data and immediately raise ConnectionEvent::LocalClose
-  FlushWriteAndDelay // Flush pending write data and delay raising a ConnectionEvent::LocalClose
-                     // until the delayed_close_timeout expires
+  FlushWriteAndDelay, // Flush pending write data and delay raising a ConnectionEvent::LocalClose
+                      // until the delayed_close_timeout expires
+  ResetNoFlush // Immediately raise ConnectionEvent::LocalClose without flushing any data and allow
```

Comment:
NoFlushAndAllowReuse maybe?
Matt, do you think belongs as another close option?  I think API concerns aside it'd be better as a socket option the connection exposes setting which in this case would be set right before closing.

Reply:
Yeah I think that would be reasonable also, though I do think that having a RST close option is something that we probably want eventually even server side in WAF/(D)DoS scenarios?

Reply:
SGTM, that sounds useful.  I mainly prefer to avoid the "no flush" option exploding into
NoFlush
NoFlushAndAllowReuse
NoFlushRstAndAllowReuse
NoFlushRst
I'm open to anything which lowers the number of combinations we end up with :-)

Reply:
How about:
```
NoFlushAndRST, // Allows immediate address reuse
```
I'll move the processing to the ConnectionImpl and remove the assert so that it can be also used for server connections.

Reply:
@mattklein123 It seems my current implementation that restricts the no linger option to the connections with source address and port set is too limiting considering potential uses in DoS mitigation scenarios. IMO i'll need to remove these restrictions and set the socket option before `close()` whenever the close type is `NoFlushAndRST`. Thoughts?

Reply:
Yes I think that makes sense, but then we have to more tightly restrict when that open is set.